### PR TITLE
Modify compaction code to be adapted to Rowset interface

### PR DESCRIPTION
**Pull request by chaoyli**

### CODE REVIEW

File: be/src/olap/cumulative_compaction.cpp [hunk 1]
```diff
@@ -462,12 +460,12 @@ OLAPStatus CumulativeCompaction::_do_cumulative_compaction() {
...
-OLAPStatus CumulativeCompaction::_update_header(vector<SegmentGroup*>* unused_indices) {
-    vector<SegmentGroup*> new_indices;
-    new_indices.push_back(_new_segment_group);
+OLAPStatus CumulativeCompaction::_update_header(vector<RowsetSharedPtr>* unused_indices) {
+    vector<RowsetSharedPtr> new_indices;
+    new_indices.push_back(_rowset);
 
     OLAPStatus res = OLAP_SUCCESS;
-    res = _tablet->replace_data_sources(&_need_merged_versions, &new_indices, unused_indices);
+    res = _tablet->modify_rowsets(&_need_merged_versions, &new_indices, unused_indices);
```
Comment (kangpinghuang) [inline be/src/olap/cumulative_compaction.cpp]: I think new_indices and unused_indices are not good names. pls change them

File: be/src/olap/cumulative_compaction.h [hunk 2]
```diff
@@ -110,13 +112,13 @@ class CumulativeCompaction {
     // 返回值：
     // - 如果成功，返回OLAP_SUCCESS
     // - 如果不成功，返回相应错误码
-    OLAPStatus _update_header(std::vector<SegmentGroup*>* unused_indices);
+    OLAPStatus _update_header(std::vector<RowsetSharedPtr>* unused_indices);
```
Comment (kangpinghuang) [inline be/src/olap/cumulative_compaction.h]: change the unused_indices name

File: be/src/olap/merger.cpp [hunk 3]
```diff
@@ -35,24 +35,24 @@ using std::vector;
...
+OLAPStatus Merger::merge(const vector<RowsetReaderSharedPtr>& rs_readers,
+                         const Version& version, uint64_t* merged_rows, uint64_t* filted_rows) {
     // Create and initiate reader for scanning and multi-merging specified
     // OLAPDatas.
     Reader reader;
     ReaderParams reader_params;
     reader_params.tablet = _tablet;
     reader_params.reader_type = _reader_type;
-    reader_params.olap_data_arr = olap_data_arr;
+    reader_params.olap_data_arr = rs_readers;
```
Comment (kangpinghuang) [inline be/src/olap/merger.cpp]: change the olap_data_arr name

File: be/src/olap/rowset/rowset.h [hunk 4]
```diff
@@ -58,6 +58,12 @@ class Rowset {
     virtual bool zero_num_rows() const = 0;
 
     virtual size_t num_rows() const = 0;
+
+    virtual Version version() const;
```
Comment (kangpinghuang) [inline be/src/olap/rowset/rowset.h]: should use virtual Version version const = 0;
change the following.

File: be/src/olap/tablet.cpp [hunk 5]
```diff
@@ -105,56 +105,101 @@ bool Tablet::can_do_compaction() {
...
-                                          vector<std::shared_ptr<RowsetReader>>* rs_readers) const {
-    DCHECK(rs_readers != NULL && rs_readers->empty());
-    for (auto version : version_vec) {
-        auto it2 = _rs_version_map.find(version);
-        if (it2 == _rs_version_map.end()) {
+OLAPStatus Tablet::capture_consistent_rowsets(const vector<Version>& version_path,
+                                        vector<RowsetSharedPtr>* rowsets) const {
+    DCHECK(rowsets != nullptr && rowsets->empty());
+    for (auto version : version_path) {
+        auto it = _rs_version_map.find(version);
```
Comment (kangpinghuang) [inline be/src/olap/tablet.cpp]: auto -> auto&?

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/base_compaction.cpp [hunk 6]
```diff
@@ -142,12 +139,20 @@ OLAPStatus BaseCompaction::run() {
...
+        return OLAP_ERR_BE_ERROR_DELETE_ACTION;
+    }
+
     VLOG(3) << "elapsed time of doing base compaction:" << stage_watch.get_elapse_time_us();
 
     // 4. make new versions visable.
     //    If success, remove files belong to old versions;
     //    If fail, gc files belong to new versions.
-    vector<SegmentGroup*> unused_olap_indices;
+    vector<RowsetSharedPtr> unused_olap_indices;
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: OK. I will change it to unused_rowsets.

File: be/src/olap/base_compaction.cpp [hunk 7]
```diff
@@ -319,22 +306,30 @@ bool BaseCompaction::_check_whether_satisfy_policy(bool is_manual_trigger,
...
+    RowsetIdGenerator::instance()->get_next_id(_tablet->data_dir(), &rowset_id);
+    RowsetBuilderContext context = {_tablet->partition_id(), _tablet->tablet_id(),
+                                    _tablet->schema_hash(), rowset_id, 
+                                    RowsetTypePB::ALPHA_ROWSET, _tablet->rowset_path_prefix(),
+                                    _tablet->tablet_schema(), _tablet->num_key_fields(),
+                                    _tablet->num_short_key_fields(), _tablet->num_rows_per_row_block(),
+                                    _tablet->compress_kind(), _tablet->bloom_filter_fpp()};
+    RowsetBuilder* builder = new AlphaRowsetBuilder(); 
+    builder->init(context);
+    if (builder == nullptr) {
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: OK

File: be/src/olap/cumulative_compaction.cpp (hunk 1 above)
Comment (chaoyli) [reply be/src/olap/cumulative_compaction.cpp]: OK

File: be/src/olap/base_compaction.cpp [hunk 8]
```diff
@@ -319,22 +306,30 @@ bool BaseCompaction::_check_whether_satisfy_policy(bool is_manual_trigger,
...
+        OLAP_LOG_WARNING("fail to new rowset.");
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: OK

File: be/src/olap/merger.cpp (hunk 3 above)
Comment (chaoyli) [reply be/src/olap/merger.cpp]: I will change it to rs_readers name.

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/base_compaction.h [hunk 9]
```diff
@@ -82,35 +84,35 @@ class BaseCompaction {
...
     // - row_count: 生成Base过程中产生的row_count
     //
     // 返回值：
     // - 如果执行成功，则返回OLAP_SUCCESS；
     // - 其它情况下，返回相应的错误码
     OLAPStatus _do_base_compaction(VersionHash new_base_version_hash,
-                                  std::vector<ColumnData*>* base_data_sources,
-                                  uint64_t* row_count);
+                                   const std::vector<RowsetSharedPtr>& rowsets,
+                                   uint64_t* row_count);
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/base_compaction.h]: It's OK

File: be/src/olap/base_compaction.h [hunk 10]
```diff
@@ -82,35 +84,35 @@ class BaseCompaction {
...
     // 输出参数：
-    // - unused_olap_indices: 需要被物理删除的SegmentGroup*
+    // - unused_rowsets: 需要被物理删除的Rowset*
     //
     // 返回值：
     // - 如果执行成功，则返回OLAP_SUCCESS；
     // - 其它情况下，返回相应的错误码
     OLAPStatus _update_header(uint64_t row_count,
-                              std::vector<SegmentGroup*>* unused_olap_indices);
+                              std::vector<RowsetSharedPtr>* unused_rowsets);
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/base_compaction.h]: It's OK

File: be/src/olap/cumulative_compaction.h [hunk 11]
```diff
@@ -27,10 +27,12 @@
 #include "olap/rowset/column_data.h"
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/cumulative_compaction.h]: OK

File: be/src/olap/merger.h [hunk 12]
```diff
@@ -29,23 +29,25 @@ class ColumnData;
 class Merger {
 public:
     // parameter index is created by caller, and it is empty.
-    Merger(TabletSharedPtr tablet, SegmentGroup* index, ReaderType type);
+    //Merger(TabletSharedPtr tablet, RowsetSharedPtr rowset, ReaderType type);
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/merger.h]: OK

File: be/src/olap/reader.cpp [hunk 13]
```diff
@@ -308,13 +308,13 @@ OLAPStatus Reader::init(const ReaderParams& read_params) {
...
 
-    res = _acquire_data_sources(read_params);
+    res = _capture_rs_readers(read_params);
     if (res != OLAP_SUCCESS) {
         OLAP_LOG_WARNING("fail to init reader when acquire data sources.[res=%d]", res);      
         return res;
     }
 
-    for (auto i_data: _data_sources) {
+    for (auto i_data : _rs_readers) {
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/reader.cpp]: This place will be changed in reader refactor.

File: be/src/olap/merger.h [hunk 14]
```diff
@@ -29,23 +29,25 @@ class ColumnData;
...
+                     const Version& version, uint64_t* merged_rows, uint64_t* filted_rows);
 
     // 获取在做merge过程中累积的行数
     uint64_t row_count() {
         return _row_count;
     }
 private:
     TabletSharedPtr _tablet;
-    SegmentGroup* _segment_group;
+    //RowsetSharedPtr _rowset;
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/merger.h]: OK

File: be/src/olap/schema_change.cpp [hunk 15]
```diff
@@ -1523,22 +1523,21 @@ OLAPStatus SchemaChangeHandler::_do_alter_tablet(
...
-        ref_tablet->acquire_data_sources_by_versions(
-                versions_to_be_changed, &olap_data_arr);
-        if (olap_data_arr.size() < 1) {
+        ref_tablet->capture_rs_readers(versions_to_be_changed, &rs_readers);
+        if (rs_readers.size() < 1) {
             OLAP_LOG_WARNING("fail to acquire all data sources."
                              "[version_num=%d data_source_num=%d]",
                              versions_to_be_changed.size(),
-                             olap_data_arr.size());
+                             rs_readers.size());
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: OK

**Review by kangpinghuang — APPROVED** [review]

**Pull request merged**